pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,594 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptables
+
+/*
+NOTE: this needs to be tested in e2e since it uses iptables for everything.
+*/
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base32"
+	"fmt"
+	"net"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/proxy"
+	"k8s.io/kubernetes/pkg/types"
+	utilexec "k8s.io/kubernetes/pkg/util/exec"
+	utiliptables "k8s.io/kubernetes/pkg/util/iptables"
+	"k8s.io/kubernetes/pkg/util/slice"
+)
+
+// NOTE: IPTABLES_MIN_VERSION is the minimum version of iptables for which we will use the Proxier
+// from this package instead of the userspace Proxier.
+// This is will not be enough, as the version number is somewhat unreliable,
+// features are backported in various distros and this could get pretty hairy.
+// However iptables-1.4.0 was released 2007-Dec-22 and appears to have every feature we use,
+// so this seems prefectly reasonable for now.
+const (
+	IPTABLES_MIN_VERSION string = "1.4.0"
+)
+
+// the services chain
+var iptablesServicesChain utiliptables.Chain = "KUBE-SERVICES"
+
+// ShouldUseIptablesProxier returns true if we should use the iptables Proxier instead of
+// the userspace Proxier.
+// This is determined by the iptables version. It may return an erorr if it fails to get the
+// itpables version without error, in which case it will also return false.
+func ShouldUseIptablesProxier() (bool, error) {
+	exec := utilexec.New()
+	minVersion, err := semver.NewVersion(IPTABLES_MIN_VERSION)
+	if err != nil {
+		return false, err
+	}
+	versionString, err := utiliptables.GetIptablesVersionString(exec)
+	if err != nil {
+		return false, err
+	}
+	versionMatcher := regexp.MustCompile("v([0-9]+\\.[0-9]+\\.[0-9]+)")
```

Comment:
Yeah was just thinking about that, I was out a for a bit myself. Back now,
will fix, rebase.

On Tue, Aug 11, 2015 at 8:54 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36817490>
> :
>
> > +
> > +// ShouldUseIptablesProxier returns true if we should use the iptables Proxier instead of
> > +// the userspace Proxier.
> > +// This is determined by the iptables version. It may return an erorr if it fails to get the
> > +// itpables version without error, in which case it will also return false.
> > +func ShouldUseIptablesProxier() (bool, error) {
> > +	exec := utilexec.New()
> > +	minVersion, err := semver.NewVersion(IPTABLES_MIN_VERSION)
> > +	if err != nil {
> > +		return false, err
> > +	}
> > +	versionString, err := utiliptables.GetIptablesVersionString(exec)
> > +	if err != nil {
> > +		return false, err
> > +	}
> > +	versionMatcher := regexp.MustCompile("v([0-9]+\\.[0-9]+\\.[0-9]+)")
>
> Naw, come on :)
>
> Make GetIptablesVersionString() return the vX.Y.Y string
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36817490>
> .
>